File: 223. Rectangle Area/Solution.py
# LeetCode #223: Rectangle Area

'''
Find the total area covered by two rectilinear rectangles in a 2D plane.

Each rectangle is defined by its bottom left corner and top right corner as shown in the figure.

Assume that the total area is never beyond the maximum possible value of int.
'''

class Solution:
    # @param {integer} A
    # @param {integer} B
    # @param {integer} C
    # @param {integer} D
    # @param {integer} E
    # @param {integer} F
    # @param {integer} G
    # @param {integer} H
    # @return {integer}
    def computeArea(self, A, B, C, D, E, F, G, H):
        area1 = (C-A) * (D-B)
        area2 = (G-E) * (H-F)
        common = max(min(C, G) - max(A, E), 0) * max(min(D, H) - max(B, F), 0)
        return area1 + area2 - common
        

File: 223. Rectangle Area/Solution.c
// LeetCode #223: Rectangle Area

/*
Find the total area covered by two rectilinear rectangles in a 2D plane.

Each rectangle is defined by its bottom left corner and top right corner as shown in the figure.

Assume that the total area is never beyond the maximum possible value of int.
*/

int max(int a, int b) {
    return a > b? a: b;
}

int min(int a, int b) {
    return a < b? a: b;
}

int computeArea(int A, int B, int C, int D, int E, int F, int G, int H) {
    int area1 = (C-A) * (D-B);
    int area2 = (G-E) * (H-F);
        
    if (E > C || A > G || B > H || F > D)
        return area1 + area2;
        
    int common = max((min(C, G) - max(A, E)), 0) * max((min(D, H) - max(B, F)), 0);
    return area1 + area2 - common;
}